### PR TITLE
Typed wrappers around parse tree nodes

### PR DESCRIPTION
**Pull request by josh11b**
These are intended to allow the structure of a parse tree node to be described more precisely in code, to support these use cases:

- Automated checking that the parse tree conforms to the expected structure. (Added to `Tree::Verify`.)
- Easier reading and understanding of the structure of the parse tree by toolchain developers. (See `parse/typed_nodes.h`.)
- Easier navigation of the parse tree, for example for tooling uses and for use when forming diagnostics.

On this last point, an object representing the file may be inspecting using `Tree::ExtractFile`, as in:
```
auto file = tree->ExtractFile();
for (AnyDeclId decl_id : file.decls) {
  // `decl_id` is convertible to a `NodeId`.
  if (std::optional<FunctionDecl> fn_decl =
      tree->ExtractAs<FunctionDecl>(decl_id)) {
    // fn_decl->params is a `TuplePatternId` (which extends `NodeId`)
    // that is guaranteed to reference a `TuplePattern`.
    std::optional<TuplePattern> params = tree->Extract(fn_decl->params);
    // `params` has a value unless there was an error in that node.
  } else if (auto class_def = tree->ExtractAs<ClassDefinition>(decl_id)) {
    // ...
  }
}
```

The `Extract...` functions collect the child nodes into the typed parse node's fields (internally using a `Tree::SiblingIterator`) for easy access. However, this is not as fast as directly observing the tree structure using the postorder strategy being used by the check stage.

These functions rely on using struct reflection on the typed parse node definitions from `parse/typed_nodes.h` to get the expected structure of child nodes and then populate them.

Note that validating these in `Tree::Verify` adds significant cost to it, and is currently included in the parsing stage. Without this change, a 10 mloc test case of lex & parse takes 4.129 s ±  0.041 s. With this change, it takes 5.768 s ±  0.036 s.

This builds upon and completes #3393.

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

### CODE REVIEW

File: toolchain/check/node_stack.h
Comment (zygoloid) [inline toolchain/check/node_stack.h:95]: (No action required.)

As future work, I think we should also look at `PopName` and `PopExpr` and think about whether we can add a `Pop<...>` taking a parse node category. Generally I think it'd be great if we base the kind of ID we get back from the node stack on the category of the parse node rather than the specific parse node to the extent that we can (eg, all expressions should result in an `InstId`).

File: toolchain/parse/extract.cpp
Comment (zygoloid) [inline toolchain/parse/extract.cpp]: Along the failure path here, I think we need to reset `it` back to where it was before we started, otherwise (eg) an `optional<tuple>` or a `SmallVector<struct>` won't work properly.

Hm. Also, the pack fold tries to extract all elements even if an earlier extraction fails. That seems OK (though perhaps wasteful) when we're not tracing, but when we are tracing it will lead to additional trace output after the point of failure, and potentially a lot of it. Maybe we should stop extracting at the first failure:

```suggestion
    // Use a fold over the `=` operator to parse fields from right to left.
    [[maybe_unused]] int unused;
    bool ok = true;
    static_cast<void>(((ok &&
                        ok = (std::get<Index>(fields) =
                              Extractable<T>::Extract(tree, it, end, trace)).has_value(),
                        unused) = ... = 0));

    if (!ok) {
      if (trace) {
        *trace << "Tuple: error\n";
      }
      return std::nullopt;
    }
```

File: toolchain/parse/extract.cpp
Comment (zygoloid) [inline toolchain/parse/extract.cpp]: Maybe include `typeid(T).name()` here?

File: toolchain/parse/extract.cpp
Comment (zygoloid) [inline toolchain/parse/extract.cpp:332]:
```suggestion
      ErrorBuilder* trace) const -> std::optional<KindName>;
```
I guess `clang-format` undoes this?

File: toolchain/parse/node_ids.h
Comment (zygoloid) [inline toolchain/parse/node_ids.h]: I think these types should have an `Id` suffix, eg `AnyDeclId`.

File: toolchain/parse/node_ids.h
Comment (zygoloid) [inline toolchain/parse/node_ids.h:92]: Would it make sense to take `const NodeKind&` here instead of `typename`? All extract wants is the kind, and passing the kind in will mean that these can be used in `typed_nodes.h` before the corresponding typed node is defined. I think this is the only remaining reason why `typed_nodes.h` needs to be in a somewhat-topological order.

File: toolchain/parse/node_kind.h
Comment (zygoloid) [inline toolchain/parse/node_kind.h]:
```suggestion
// A definition of a parse node kind. This is a NodeKind value, plus
// ancillary data such as the name to use for the node kind in LLVM IR. These
// are not copyable, and only one instance of this type is expected to exist per
// parse node kind, specifically `TypedNode::Kind`. Use `NodeKind` instead as a
// thin wrapper around a parse node kind index.
```

File: toolchain/parse/tree.cpp
Comment (zygoloid) [inline toolchain/parse/tree.cpp:253]: Can you measure any performance impact from this additional verify step? (I don't know that we have a good benchmark here, but maybe you could generate a large input and run it through `carbon compile`.)

File: toolchain/parse/typed_nodes.h
Comment (zygoloid) [inline toolchain/parse/typed_nodes.h]:
```suggestion
// `library` in `package` or `import`.
```

File: toolchain/parse/typed_nodes.h
Comment (zygoloid) [inline toolchain/parse/typed_nodes.h]:
```suggestion
// `library` as directive.
```

File: toolchain/parse/typed_nodes.h
Comment (zygoloid) [inline toolchain/parse/typed_nodes.h]: Do you want this answered before the PR lands? If not:
```suggestion
// TODO: Should this be a template?
```
I don't really have a strong opinion here. It's not much duplication compared to the overhead of writing a template and two aliases. I'd be happy to say that we don't make this into a template until / unless we want to write generic code to cover both cases at once (eg, in check).

File: toolchain/parse/typed_nodes.h
Comment (zygoloid) [inline toolchain/parse/typed_nodes.h]:
```suggestion
// Struct literals, such as `{.a = 0}`.
```

File: toolchain/parse/typed_nodes.h
Comment (zygoloid) [inline toolchain/parse/typed_nodes.h]:
```suggestion
// Struct type literals, such as `{.a: i32}`.
```

File: toolchain/parse/typed_nodes.h
Comment (zygoloid) [inline toolchain/parse/typed_nodes.h]: Probably this was originally my phrasing, but I think this is clearer the other way around:
```suggestion
// Define `Foo` as the node type for the ID type `FooId`.
```

File: toolchain/parse/typed_nodes.h
Comment (zygoloid) [inline toolchain/parse/typed_nodes.h:70]: I think it would be helpful to do something to make these subheadings stand out a bit more in the file. Given that we use markdown-inspired syntax in our comments already, how about one of these:
```suggestion
// ### Error nodes
```
or
```suggestion
// Error nodes
// -----------
```

File: toolchain/parse/typed_nodes.h
Comment (zygoloid) [inline toolchain/parse/typed_nodes.h:16]: A subheading comment before these three non-parse-node-kind helpers would be helpful.

File: toolchain/parse/typed_nodes_test.cpp
Comment (zygoloid) [inline toolchain/parse/typed_nodes_test.cpp]: I think we should rename `File::Make` to `File::Extract` for symmetry with other extraction operations. I also wonder if we can switch this around to being `tree->ExtractFile()`, but perhaps the dependencies there are awkward?

File: toolchain/parse/typed_nodes_test.cpp
Comment (zygoloid) [inline toolchain/parse/typed_nodes_test.cpp]: I wonder if we should make the type optional when the argument is a `NodeIdForKind` already, so we can write this as just `tree->IsValid(file.start)`. (Even if so, I'm happy for that to be left to a future PR!)

**Comment by chandlerc** [comment]
Could you add a PR description that details a bit more of what the end result here is, what the API looks like, etc. This is a large and complex PR, and so it is quite hard to look at it and really understand the effect it has as opposed to the implementation strategy it leverages.

File: toolchain/parse/typed_nodes_test.cpp
Comment (zygoloid) [inline toolchain/parse/typed_nodes_test.cpp:91]: It'd be nice to have an example with multiple modifiers so we can check that we extract them in the right order (reversed from sibling order).

File: toolchain/check/node_stack.h
Comment (josh11b) [reply toolchain/check/node_stack.h:95]: Added a TODO. Do note though that some parse node kinds are in multiple categories: `var` and `let` are both declarations and statements.

File: toolchain/parse/extract.cpp
Comment (josh11b) [reply toolchain/parse/extract.cpp]: Good idea.

File: toolchain/parse/node_ids.h
Comment (josh11b) [reply toolchain/parse/node_ids.h]: Done.

File: toolchain/parse/node_ids.h
Comment (josh11b) [reply toolchain/parse/node_ids.h:92]: It would make the use of this more noisy: either you write `Foo::Kind` (which only works after `Foo` is defined anyway) or you write `NodeKind::Foo` which is pretty long. I can still do it, but I really haven't had any trouble coming up with a logical order to write typed_nodes.h in that respects the dependency edges. `NodeIdNot` specifically is only used in one place to say "not the introducer", which of course was already defined as is natural.

File: toolchain/parse/typed_nodes_test.cpp
Comment (josh11b) [reply toolchain/parse/typed_nodes_test.cpp]: I took a stab at it and it seemed to have worked.

File: toolchain/parse/typed_nodes_test.cpp
Comment (josh11b) [reply toolchain/parse/typed_nodes_test.cpp:91]: Done.

File: toolchain/parse/extract.cpp
Comment (josh11b) [reply toolchain/parse/extract.cpp:332]: Indeed our autoformating undoes this.

File: toolchain/parse/typed_nodes_test.cpp
Comment (josh11b) [reply toolchain/parse/typed_nodes_test.cpp]: I split it into pieces, which worked with a forward declaration.

File: toolchain/parse/tree.cpp
Comment (josh11b) [reply toolchain/parse/tree.cpp:253]: Added a comment and to the PR description.

File: toolchain/parse/typed_nodes.h
Comment (josh11b) [reply toolchain/parse/typed_nodes.h:16]: Done.

File: toolchain/parse/typed_nodes.h
Comment (josh11b) [reply toolchain/parse/typed_nodes.h]: Updated comment.

File: toolchain/parse/typed_nodes.h
Comment (josh11b) [reply toolchain/parse/typed_nodes.h:70]: Used the latter

**Comment by josh11b** [comment]
> Could you add a PR description that details a bit more of what the end result here is, what the API looks like, etc. This is a large and complex PR, and so it is quite hard to look at it and really understand the effect it has as opposed to the implementation strategy it leverages.

Comments and PR description updated.

File: toolchain/check/node_stack.h
Comment (zygoloid) [inline toolchain/check/node_stack.h]:
```suggestion
  // TODO: Can we add a `Pop<...>` that takes a parse node category? See
```

File: toolchain/parse/tree.h
Comment (zygoloid) [inline toolchain/parse/tree.h]:
```suggestion
  auto ExtractFile() const -> File;
```

File: toolchain/parse/tree.h
Comment (zygoloid) [inline toolchain/parse/tree.h]: Would something like this make sense?
```suggestion
  // Like ExtractAs(), but malformed tree errors are not fatal. Should only be used
  // by `Verify()`.
```

File: toolchain/parse/node_ids.h
Comment (zygoloid) [reply toolchain/parse/node_ids.h:92]: OK, let's leave as-is for now.

**Review by zygoloid — APPROVED** [review]
LGTM, please also check with @chandlerc.

File: toolchain/parse/typed_nodes.h
Comment (chandlerc) [inline toolchain/parse/typed_nodes.h]:
```suggestion
// `static constexpr Kind` member initialized by calling `Define` on the
```

File: toolchain/parse/typed_nodes.h
Comment (chandlerc) [inline toolchain/parse/typed_nodes.h]:
```suggestion
// corresponding `NodeKind`, and passing in the `NodeCategory` of that kind.
// This will both associate the category with the node kind and create the
// necessary kind object for the typed node.
```

File: toolchain/parse/typed_nodes.h
Comment (chandlerc) [inline toolchain/parse/typed_nodes.h]:
```suggestion
// Each node kind (in node_kind.def) should have a corresponding type defined here
// which describes the expected child structure of that parse node.
//
// Each of these types should start with a
```

File: toolchain/parse/typed_nodes.h
Comment (chandlerc) [inline toolchain/parse/typed_nodes.h]:
```suggestion
// This should be followed by field declarations that describe the child nodes, in order, that occur in the parse tree. The `Extract...` functions on the parse tree use struct reflection on these fields to guide the extraction of the child nodes from the tree into an object of this type with these fields for convenient access.
//
// The types of these fields are special and describe the specific child node structure of the parse node. Many of these types are defined in `node_ids.h`.
//
// Valid primitive types here are:
// - `NodeId` to match any single child node
// - `FooId` to require that child to have kind `NodeKind::Foo`
// - `AnyCatId` to require that child to have a kind in category `Cat`
// - `NodeIdOneOf<A, B>` to require the child to have kind `NodeKind::A` or `NodeKind::B`
// - `NodeIdNot<A>` to match any single child whose kind is not `NodeKind::A`
//
// There a few, restricted composite field types allowed that compose types in various ways, where all of the `T`s and `U`s below are themselves valid field types:
// - `llvm::SmallVector<T>` to match any number of children matching `T`
// - `std::optional<T>` to match 0 or 1 children matching `T`
// - `std::tuple<T...>` to match children matching `T...`
// - Any provided `Aggregate` type that is a simple aggregate type such as `struct Aggregate { T x; U y; }`,
//   to match children with types `T` and `U`.
```

File: toolchain/parse/typed_nodes.h
Comment (chandlerc) [inline toolchain/parse/typed_nodes.h]: Could you add a blank line after the kind and the start of the fields? (Not suggesting the edit to all of them, but I can if needed.)

```suggestion
  static constexpr auto Kind =
      NodeKind::InvalidParseSubtree.Define(NodeCategory::Decl);
      
  InvalidParseStartId start;
  llvm::SmallVector<NodeIdNot<InvalidParseStart>> extra;
```

File: toolchain/parse/extract.cpp
Comment (chandlerc) [inline toolchain/parse/extract.cpp]: While I love the Carbon code in here, I think maybe it's better to use a more direct C++ example.

```suggestion
// ```cpp
// template<>
// struct Extractable<T> {
//   // Extract a value of this type from the sequence of nodes starting at
//   // `it`, and increment `it` past this type. Returns `std::nullopt` if
//   // the tree is malformed. If `trace != nullptr`, writes what actions
//   // were taken to `*trace`.
//   static auto Extract(Tree* tree, Tree::SiblingIterator& it,
//                       Tree::SiblingIterator end,
//                       ErrorBuilder* trace) -> std::optional<T>;
// };
// ```
```

Comment (chandlerc) [reply toolchain/parse/extract.cpp]: Honestly, not sure its an improvement... the Cabron feature here is really nice for illustrating what is going on...

So consider this optional.

File: toolchain/parse/extract.cpp
Comment (chandlerc) [inline toolchain/parse/extract.cpp]:
```suggestion
// Extract a `NodeId` as a single child.
```

File: toolchain/parse/extract.cpp
Comment (chandlerc) [inline toolchain/parse/extract.cpp]:
```suggestion
// Extract a `NodeIdInCategory<Category>` as a single child.
```

File: toolchain/parse/extract.cpp
Comment (chandlerc) [inline toolchain/parse/extract.cpp]: Note: I have zero opinion on how best to match articles, and for `std` in particular, there's no universal pronunciation. I asked @zygoloid to give me his intended article given the different letters following and have suggested those here. Consider all of this optional. =]

```suggestion
// Extract an `llvm::SmallVector<T>` by extracting `T`s until we can't.
```

File: toolchain/check/node_stack.h
Comment (chandlerc) [inline toolchain/check/node_stack.h]:
```suggestion
  // TODO: Return a `Parse::NodeIdForKind<RequiredParseKind>` instead.
```

File: toolchain/check/check.cpp
Comment (chandlerc) [inline toolchain/check/check.cpp]:
```suggestion
      // TODO: Switch to `Parse::Name##Id(parse_node)` here.
```

File: toolchain/parse/tree.h
Comment (chandlerc) [inline toolchain/parse/tree.h]: What do you think about calling this `VerifyExtractAs` (or something similar)? I worry about the name `TryExtractAs` seeming like it should be used to try one of two different node types to see which one works, and it sounds like that's not really the use case here.

```suggestion
  // Like ExtractAs(), but malformed tree errors are not fatal. Should only be
  // used by `Verify()`.
  template <typename T>
  auto VerifyExtractAs(NodeId node_id, ErrorBuilder* trace) const
      -> std::optional<T>;
```

Another idea, but this one entirely optional: would returning `ErrorOr<T>` make sense here, to parallel `Verify` itself?

File: toolchain/parse/tree.h
Comment (chandlerc) [inline toolchain/parse/tree.h]: Could we move this below the non-verify extract APIs, but keep it with the verifying extract API?

**Review by chandlerc — COMMENTED** [review]
Just wanted to say, this is really cool. I love that it gives us a structural verification layer that we can both enforce and leverage in places where it makes sense.

A bunch of comments & suggestions inline just trying to help make sure the code comments make sense to a reader with less context on everything. (Many were paired with Josh and/or Richard.)

Some are super nit-picky, sorry for that, just saw them as I was reading and figured I would suggest edits. None of those are important if they get in the way of merges or anything. And I've tried to mark a few other places where things are just optional suggestions.

File: toolchain/parse/tree.h
Comment (josh11b) [reply toolchain/parse/tree.h]: Issue with `ErrorOr<T>` is the one user includes more context in the error.

File: toolchain/parse/tree.h
Comment (josh11b) [reply toolchain/parse/tree.h]: Done.

**Comment by chandlerc** [comment]
Awesome, mashing the merge button!

**Review by chandlerc — APPROVED** [review]

**Pull request merged**
